VA-Explorer/va_explorer
# VA Explorer 148: Support a deletion workflow for duplicate VAs

# Description
This pull request addressed [VA Explorer 148](https://github.com/VA-Explorer/va_explorer/issues/148), Support a deletion workflow for duplicate VAs. 

### Configuration
Implementers of VA Explorer will identify a subset of key VA fields that will, in concert, uniquely identify a VA if they want to use this feature. The fields,`QUESTIONS_TO_AUTODETECT_DUPLICATES`, can be passed into the application as an environment variable. The default value for this variable is `None`.

### Loading New VAs
When this feature is turned on and new VAs are imported, we populate the field `unique_va_identifier`. This field is the md5 hash of the concatenation of the fields in `QUESTIONS_TO_AUTODETECT_DUPLICATES`. After `unique_va_identifier` is populated, we mark VAs as duplicate if their `unique_va_identifier` fields match. The oldest VA (by created) amongst a set of VAs with matching `unique_va_identifiers` is designated as the non-duplicate. 

### Managing Duplicate VAs
The system surfaces potentially duplicate VAs in the user interface in a separate Data Cleanup tab. The system offers 5 permissions related to Data Cleanup. All permissions are added to the Admin User group, only, in the initialize_groups command. 
- **view_datacleanup:** View the Data Cleanup index page
- **download:** Download an individual duplicate Verbal Autopsy
- **bulk_download:** Download all duplicate Verbal Autopsies
- **delete_verbalautopsy:** Delete an individual duplicate Verbal Autopsy
- **bulk_delete:** Delete all duplicate Verbal Autopsies

Users can manage duplicate Verbal Autopsies by deleting them or editing them to remove them from detection. 

- Deletion: When a user deletes a duplicate Verbal Autopsy, the system will perform a "soft delete" of the record, where we set the timestamp, deleted_at, in the database and hide the VA from the user interface  The soft delete will prevent the “deleted” VA from being reimported because we currently check for the presence of the same instanceid in the db when importing new records. Since the instanceid will still be in the db, it will not be re-imported. The soft delete approach will allow us to surface “deleted” VAs in a recycling bin view if the user wants to recover them (not implemented in this PR). 
- Edit: If a user edits any of the fields in `QUESTIONS_TO_AUTODETECT_DUPLICATES`, the system repopulates `unique_va_identifier` and re-identifies duplicates 

### Inspecting Duplicate VAs
Administrators can inspect duplicate VAs by downloading all duplicate VAs in the system or downloading individual duplicate VAs to CSV. When the latter action is chosen, all VAs matching that VA are present in the CSV (including the non-duplicate VA -- this appears as the first VA in the row)

### Marking Existing VAs as Duplicate
To mark existing VAs as duplicate, this PR adds a command, `mark_vas_as_duplicate`. This command should be run prior to using the feature in the User Interface to avoid unexpected results. 

## Important Changes
`config/settings/base.py`
- Adds `QUESTIONS_TO_AUTODETECT_DUPLICATES` configuration

`va_explorer/models.py`
- Adds generic SoftDeletionModel and SoftDeletionManager

`va_explorer/utils/file_io.py`
- Adds utility method to download a queryset as a CSV

`va_explorer/va_data_cleanup/apps.py`
- Adds VaDataCleanup app

`va_explorer/va_data_cleanup/views.py`
- Adds the following views: DataCleanupIndexView, DownloadIndividual, DownloadAll, DownloadQuestions

`va_explorer/va_data_management/management/commands/mark_vas_as_duplicate.py`
- Adds command for marking pre-existing Verbal Autopsies as duplicate 

`va_explorer/va_data_management/models.py`
- Adds methods to mark VAs as duplicates and handle updating of duplicates 
- Adds the following managers: DhisStatusManager, CauseOfDeathManager, CauseCodingIssueManager 

`va_explorer/va_data_management/utils/loading.py`
- Adds unique_va_identifier to the VA
- Marks VAs as duplicate

`va_explorer/va_data_management/views.py`
- Adds the following views: Delete and DeleteAll 

## Testing Recommendations
1. Run `./manage.py mark_vas_as_duplicate` without running the migration. Confirm error message printed to console.
2. Run the migrations. 
3. Run `./manage.py initialize_groups` to add the permissions noted above to the Admin User. 
4. Run `./manage.py mark_vas_as_duplicate` without adding `QUESTIONS_TO_AUTODETECT_DUPLICATES` to your .env file. Confirm error message printed to the console.
5. Add `QUESTIONS_TO_AUTODETECT_DUPLICATES` to your .env file. The value for this variable must be a comma-separated string (list did not work with Docker). Run `./manage.py mark_vas_as_duplicate`. Confirm that the `unique_va_identifier` and `duplicate` fields are populated appropriately. See example below for .env file configuration: 
```
QUESTIONS_TO_AUTODETECT_DUPLICATES="Id10017, Id10018, Id10019, Id10020, Id10021, Id10022, Id10023"
```
6. Run `./manage.py load_va_csv <NAME OF CSV>`. Confirm that the `unique_va_identifier` and `duplicate` fields are populated appropriately.
7. While logged in as an Admin User, view the Data Cleanup page, /va_data_cleanup. Confirm that all of the actions on the page work as expected. Confirm that the language on the page accurately explains the feature. Confirm that the tooltips work on the Download and Delete columns in the Individual Actions table. Confirm that the language in the tooltips makes sense. 
8. While logged in as an Admin User, go to the Data Cleanup page and delete an individual VA. Confirm that after deleting a Verbal Autopsy, it no longer appears in the Data Management tab and that the counts on the Home page and Dashboard are updated. 
9. While logged in as an Admin User, go to the Data Cleanup page, /va_data_cleanup. View one of the duplicate verbal autopsies and then edit the VA, being sure to edit the VA in such a way that it is no longer duplicate. After editing, confirm that the VA no longer appears on the Data Cleanup page. Confirm that the duplicate VA warning banner is no longer present on the VA's show page. 
10. While logged in as an Admin User, go to the Data Management page. View one of the verbal autopsies and then edit the VA, being sure to edit the VA in such a way that it is now a duplicate. Confirm that the duplicate VA warning banner is present on the VA's show page. Confirm that the VA appears on the Data Cleanup page. 
11. While logged in as a non-Admin User, confirm that permission denied for the Data Cleanup page. Confirm that the Data Cleanup menu item is absent from the navigation bar. Confirm that permission denied for downloading, bulk downloading, deleting, and bulk deleting. 

### Next Steps/Questions
- This PR does not require the user to take any action prior to editing a duplicate VA. Adam suggested that duplicate VAs be "locked" and we require the user to unlock the VA first. As this PR is already (sadly) quite long, I did not tackle this yet.
- This PR does not offer a recycling bin view to recover deleted VAs. Because we are soft deleting VAs, this could be added if needed.
- The permissions added may be too granular- thoughts?
- Do we need to expose the audit trail for deletions that occur in the application? 
- It might be nice to mark the questions that are used to identify duplicates with a symbol. Currently, the user does have the ability to download a list of the questions from the Data Cleanup page.
